CLUE_Rock_Paper_Scissors/advanced/rps_comms.py

```diff
@@ -0,0 +1,361 @@
+                        send_ad,
+                        *receive_ads_types,
+                        scan_time=DEF_SEND_TIME_S,
+                        ad_interval=MIN_AD_INTERVAL,
+                        buffer_size=1800,
+                        minimum_rssi=-90,
+                        receive_n=0,
+                        seq_tx=None,
+                        match_locally=True,
+                        scan_response_request=False,
+                        ad_cb=None,
+                        ads_by_addr={},
+                        names_by_addr={},
+                        name_cb=None,
+                        endscan_cb=None
+                        ):
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    """Send an Advertisement send_ad and then wait for up to scan_time to
+       receive receive_n Advertisement packets from other devices.
+       If receive_n is 0 then wait for the remaining scan_time.
+       Returns list of received Advertisements not necessarily in arrival order and
+       dictionary indexed by the compressed text representation of the address with a list
+       of tuples of (advertisement, bytes(advertisement)).
+       This MODIFIES send_ad by setting sequence_number and ack if those
+       properties are present.
+       This is likely to run for a fraction of second longer than scan_time.
+       The default scan_response_request of False should reduce traffic and
+       may reduce collisions.
+       The buffer_size of 1800 helps to prevent 1784 MemoryError
+       from dict enlargement including the interpreter's symbol table.
+       """
+
+    sequence_number = None
+    if seq_tx is not None and hasattr(send_ad, "sequence_number"):
+        sequence_number = seq_tx[0]
+        send_ad.sequence_number = sequence_number
+        seq_tx[0] += 1
+
+    # A dict to store unique Advertisement indexed by mac address
+    # as text string
+    cls_send_ad = type(send_ad)
+    received_ads_by_addr = dict(ads_by_addr)  # Will not be a deep copy
+    if receive_ads_types:
+        rx_ad_classes = receive_ads_types
+    else:
+        rx_ad_classes = (cls_send_ad,)
+
+    if match_locally:
+        ss_rx_ad_classes = (Advertisement,)
+    else:
+        ss_rx_ad_classes = rx_ad_classes
+
+    blenames_by_addr = dict(names_by_addr)  # Will not be a deep copy
+
+    # Look for packets already received of the cls_send_ad class (type)
+    send_ad_rxs = {}
+    # And make a list of sequence numbers already acknowledged
+    acks = {}
+    for addr_text, adsnb_per_addr in received_ads_by_addr.items():
+        if cls_send_ad in [type(andb[0]) for andb in adsnb_per_addr]:
+            send_ad_rxs[addr_text] = True
+
+        # Pick out any Advertisements with an ack field with a value
+        acks_thisaddr = list(filter(lambda adnb: hasattr(adnb[0], "ack")
+                                    and isinstance(adnb[0].ack, int),
+                                    adsnb_per_addr))
```

Comment:
I believe this could be a list comprehension with an `if`; might be a little easier to read.

Reply:
Done